### ISSUE TITLE
Support vsync in opengl windows

### ISSUE THREAD

**Issue by gregory38** [PCSX2/pcsx2#529]
It must be easy to do.

http://stackoverflow.com/questions/589064/how-to-enable-vertical-sync-in-opengl

https://www.opengl.org/wiki/Swap_Interval

```
    wglSwapIntervalEXT = GetProcAddress("wglSwapIntervalEXT");
```

**Comment by ssakash** [comment]
wglSwapIntervalExt function is Windows only.

**Comment by gregory38** [comment]
??? Did you read tag and title?

**Comment by ssakash** [comment]
sorry, I overlooked both of those things. :p 

by the way, should we swap interval with glFinish (or) without glFinish?

**Comment by gregory38** [comment]
For future

```
[22:30:33] <Armada> MrKB: if you specify the WS_POPUPWINDOW style you can bypass the window manager in OpenGL basically for free
[22:31:19] <Armada> relies on driver heuristics, so it's undocumented, but it's pretty consistent
[22:31:30] <Armada> it's what I use for exclusive mode at Dolphin
```

**Comment by karasuhebi** [comment]
Vsync isn't working when using OpenGL on Windows. I see tearing. Can anyone confirm?

**Comment by avih** [comment]
> Vsync isn't working when using OpenGL on Windows. I see tearing. Can anyone confirm?

I didn't try, but there are few factors in play here:

TL;DR, OpenGL in Windows in general:
- in a window, no tearing should happen but vsync timing is not likely to work.
- In full screen, tearing may or may not happen, and vsync timing might or might not work (respectively, if it can tear then vsync timing CAN work) - depending on the driver and the the program (PCSX2 here).
- With PCSX2, even in full screen, it's unlikely to have tearing, and vsync timing is not likely to not work.

When displaying content in a window (i.e. not full screen), Windows should not be able to display tearing even if you try really hard. Or at least it shouldn't when the compositor ("Aero") is being used, which is always on Windows 8 onward, and when using the default theme with windows Vista and 7.

Because the desktop and all windows are vsync'ed implicitly via the compositor, there's not really a monitor vsync signal to sync to, but rather a compositing software which can pick the output buffer from the OpenGL driver at potentially irregular times.

Whether or not the OpenGL driver can sync to this potentially irregular timing for a window when used with `wglSwapIntervalEXT(1)` ("vsync") is up to driver implementation.

Full screen is a bit more complex. The rule of thumb should be that when the content passes through the compositor then it's like when not in full screen, and when it doesn't go through the compositor then tearing can happen and vsync timing should work.

The issue is that the compositor might or might not be used in full screen, depending on few factors:
- Exclusive full screen OpenGL window - doesn't go through the compositor.
- Non exclusive full screen OpenGL window (active, normal but borderless window at the size of the monitor) - technically should go through the compositor, but some drivers may recognize this case and make it exclusive anyway (Intel driver does that, not sure about others). You can recognize when the driver goes exclusive anyway when it looks like the window becomes always on top (but it does become non exclusive after ALT-Tab for instance).
- Window or full screen on non-primary monitor - don't even go there, it's a mess as far as compositing/timing goes (AFAIK it always go through the compositor, and the compositor timing on a secondary monitor is still tied to the compositor timing of the primary monitor).

In PCSX2, it's even more complex, even when using full screen. Not only does PCSX2 not try to use exclusive full screen, but it also puts the OpenGL GS window inside another window (the `GSFrame` class window), which probably prevents the driver from turning it into exclusive mode automatically.

So overall, it's non trivial. If we wanted to use OpenGL vsync, then we can only do that in full screen, and on this case we'll probably at least have to take the "internal" OpenGL GS window out of `GSFrame` and make it cover the monitor exactly - to allow some drivers to try to make it exclusive.

**Comment by karasuhebi** [comment]
Thank you very much for that post @avih, it was very detailed and insightful. I appreciate it!

Just in case I can be in any way helpful with this problem, I'm going to describe as much as I can about my situation and the observations I made:

Last night when I sent that message about Vsync not working, I had my computer connected to the TV in my living room using a DisplayPort-to-HDMI active adapter (it needs to be active since my two TMDS clocks are being used by my DVI-D and HDMI monitors respectively) and a 10' HDMI cable (along with a 6' HDMI extension). I had both of my monitors disabled and made my TV the primary display. The game I saw the tearing on was Final Fantasy X and I noticed it during use of the OpenGL renderer in fullscreen. I honestly didn't have the game in a window for long enough to see if there was any tearing or not, but I just did another quick test using my main HDMI monitor and I still see tearing in OGL fullscreen but I don't see it at all when it's in a window.

So it seems what's going on here is basically exactly what you described in your second point: Even though PCSX2's OGL fullscreen is actually just a borderless window (which should go through the compositor and be vsynced), my GPU's driver decided to make it exclusive fullscreen instead. I'm guessing the behavior you described in your second-to-last paragraph about butting the OGL GS window inside another window is the reason why even though the window is being made exclusive, I still see tearing. This is just a guess though.

>So overall, it's non trivial. If we wanted to use OpenGL vsync, then we can only do that in full screen

So should the vsync setting under the GS Window tab in Emulation Settings be changed so people know that only applies vsync to the D3D renderer?

**Comment by mirh** [comment]
> It must be easy to do.

It isn't 😚 
Swapbuffers in a composited environment incurs into a smoothness [penalty](https://www.opengl.org/pipeline/article/vol003_7/). 
Moreover, if [gfwl](https://github.com/glfw/glfw/issues/24#issuecomment-104452257) guys aren't dumb, [at least under Windows DWM] anything but SwapIntervalEXT(0) will cause conflicts with its own syncing.
